// nativescript-auth0/auth0.ios.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var common = require("./auth0.common");
var appSettings = require("application-settings");
var frameModule = require("ui/frame");
var Auth0Lock = (function (_super) {
    __extends(Auth0Lock, _super);
    function Auth0Lock(options) {
        return _super.call(this, options) || this;
    }
    Auth0Lock.prototype.show = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            var page = frameModule.topmost().ios.controller;
            var lock = Lock.classic();
            if (_this.options.scope) {
                var scopeItems = _this.options.scope.join(" ");
                console.log("Adding scope of " + scopeItems);
            }
            lock.onAuthWithCallback(function (credientials) {
                console.log("Authentication Success");
                var creds = {
                    accessToken: credientials.accessToken,
                    idToken: credientials.idToken,
                    refreshToken: credientials.refreshToken,
                };
                appSettings.setString(common.Auth0Lock._tokenKey, JSON.stringify(creds));
                resolve({
                    credentials: credientials
                });
            });
            lock.presentFrom(page);
            console.log("PRESENT");
        });
    };
    return Auth0Lock;
}(common.Auth0Lock));
exports.Auth0Lock = Auth0Lock;
